param/parameterized.py

```diff
@@ -261,29 +262,64 @@ def _f(self, obj, val):
 @accept_arguments
 def depends(func, *dependencies, **kw):
     """
-    Annotates a Parameterized method to express its dependencies.
-    The specified dependencies can be either be Parameters of this
-    class, or Parameters of subobjects (Parameterized objects that
-    are values of this object's parameters).  Dependencies can either
-    be on Parameter values, or on other metadata about the Parameter.
+    Annotates a function or Parameterized method to express its
+    dependencies.  The specified dependencies can be either be
+    Parameter instances or if a method is supplied they can be
+    defined as strings referring to Parameters of the class,
+    or Parameters of subobjects (Parameterized objects that are
+    values of this object's parameters).  Dependencies can either be
+    on Parameter values, or on other metadata about the Parameter.
     """
 
     # python3 would allow kw-only args
     # (i.e. "func,*dependencies,watch=False" rather than **kw and the check below)
     watch = kw.pop("watch",False)
-    assert len(kw)==0, "@depends accepts only 'watch' kw"
-
-    # TODO: rename dinfo
-    _dinfo = getattr(func, '_dinfo', {})
-    _dinfo.update({'dependencies': dependencies,
-                   'watch': watch})
 
     @wraps(func)
     def _depends(*args,**kw):
         return func(*args,**kw)
 
-    # storing here risks it being tricky to find if other libraries
-    # mess around with methods
+    deps = list(dependencies)+list(kw.values())
+    string_specs = False
+    for dep in deps:
+        if isinstance(dep, basestring):
+            string_specs = True
+        elif not isinstance(dep, Parameter):
+            raise ValueError('Reactive functions must have '
+                             'only parameters as dependencies. '
+                             'Found %s type instead.' %
+                             type(dep).__name__)
+        elif not (isinstance(dep.owner, Parameterized) or
+                  (isinstance(dep.owner, ParameterizedMetaclass))):
+            owner = 'None' if dep.owner is None else '%s class' % type(dep.owner).__name__
+            raise ValueError('Reactive functions input parameters '
+                             'must be associated with a Parameterized '
+                             'instance, parameter owner is %s.' % owner)
+
+    if len({type(dep) for dep in deps}) > 1:
+        raise ValueError('Dependencies must either be defined as strings '
+                         'referencing parameters on the class defining '
+                         'the decorated method or as parameter instances, '
```

Comment:
Full stop instead of a comma here and a new sentence for 'Mixing of string specs and parameter instances is not supported.'?